src/Microsoft.AspNetCore.Dispatcher/DispatcherMiddleware.cs

```diff
@@ -1,60 +1,93 @@
+        public DispatcherMiddleware(IOptions<DispatcherOptions> options, RequestDelegate next)
         {
+            _options = options.Value;
             _next = next;
         }
 
         public async Task Invoke(HttpContext httpContext)
         {
-            var dictionary = new Dictionary<string, DispatcherFeature>
+            foreach (var entry in _options.DispatcherEntryList)
             {
+                var parsedTemplate = entry.RouteTemplate;
+                var defaults = GetDefaults(parsedTemplate);
+                var templateMatcher = new TemplateMatcher(parsedTemplate, defaults);
+                var values = new RouteValueDictionary();
+
+                foreach (var endpoint in entry.Endpoints)
                 {
-                    "/example",
-                    new DispatcherFeature
+                    if (templateMatcher.TryMatch(httpContext.Request.Path, values))
+                    {
+                        if (MatchURLToEndpoint(values, endpoint.RouteValueDictionary))
                         {
-                            Endpoint = new DispatcherEndpoint("example"),
-                            RequestDelegate = async (context) =>
+                            var dispatcherFeature = new DispatcherFeature
                             {
-                                await context.Response.WriteAsync("Hello from the example!");
-                            }
+                                Endpoint = endpoint,
+                                RequestDelegate = endpoint.RequestDelegate
+                            };
+
+                            httpContext.Features.Set<IDispatcherFeature>(dispatcherFeature);
                         }
-                },
+                    }
+                }
+            }
+
+            await _next(httpContext);
+        }
+
+        private RouteValueDictionary GetDefaults(RouteTemplate parsedTemplate)
+        {
+            var result = new RouteValueDictionary();
+
+            foreach (var parameter in parsedTemplate.Parameters)
+            {
+                if (parameter.DefaultValue != null)
                 {
-                    "/example2",
-                    new DispatcherFeature
-                        {
-                            Endpoint = new DispatcherEndpoint("example2"),
-                            RequestDelegate = async (context) =>
-                            {
-                                await context.Response.WriteAsync("Hello from the second example!");
-                            }
-                        }
-                },
-            };
+                    result.Add(parameter.Name, parameter.DefaultValue);
+                }
+            }
+
+            return result;
+        }
+
+        private bool MatchURLToEndpoint(RouteValueDictionary currentURLRouteValueDictionary, RouteValueDictionary endpointRouteValueDictionary)
+        {
+            var endpointMatch = new RouteValueDictionary();
```

Comment:
This isn't use for anything other than maintaining a count, so we just use an int if you need that